Add mobile data to WCTracker

joshbetz opened the pull request:
Adds a telemetry endpoint to record WCTracker data from the apps. We track each platform independently, so we can see which versions of the apps are used on a given site.

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adds a telemetry endpoint that we can use from the mobile apps to add mobile data usage to WCTracker.

### How to test the changes in this Pull Request:

1. Send a POST request to the `/wp-json/wc-telemetry/tracker` endpoint with body: `{ "platform": "ios", "version": "1.2" }`
    * Example (requires [the basic auth plugin](https://github.com/WP-API/Basic-Auth)): `curl -u user:password -X POST -H "Content-Type: application/json" -d '{ "platform": "ios", "version": "1.2" }' http://localhost/wp-json/wc-telemetry/tracker`
2. Via WP-CLI or similar, execute `WC_Tracker::get_tracking_data()`. Observe that `wc_mobile_usage` shows the correct platform and version.
    * Example:  `wp eval 'var_dump(WC_Tracker::get_tracking_data());'`
3. Send another POST request to the telemetry endpoint with body: `{ "platform": "ios", "version": "1.3" }`
4. Observe that `wc_mobile_usage` has recorded the updated version correctly.
5. Send another POST request to the telemetry endpoint with the original body: `{ "platform": "ios", "version": "1.2" }`
6. Observe that `wc_mobile_usage` has NOT updated with the older version number.

Note: If using `wp shell` to test, you may have to close the shell session and re-open if the option value is cached.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Dev - Add mobile data to WCTracker

### FOR PR REVIEWER ONLY:

* [x] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshbetz committed:
Add mobile data to WCTracker

Adds an endpoint to consolidate the data that is needed every time the
app is opened instead of relying on multiple endpoints.

This is also a good place to include some basic mobile data usage in
WCTracker. We track each platform independently, so we can see which
versions of the apps are used on a given site.

joshbetz committed:
Convert to /telemetry endpoint

Instead of overloading the application startup endpoint with recording
telemetry data, we'll start with a dedicated telemetry endpoint. This
should be easier to reason about and the additional overhead of a
separate endpoint is minimal.

joshbetz committed:
Convert to POST endpoint

We're only writing data now, so it's more natural for this to be a POST
endpoint. We'll send the data in the request body instead of as query
parameters.

joshbetz committed:
Use gmdate instead of date per coding standards

gmdate always returns a GMT date, so is not suspectible to runtime
timezone changes

joshbetz committed:
Fix function name

We were calling this function with a different name

joshbetz committed:
Add argument validation on telemetry endpoint

Properly validate the platform and version parameters in the request
body for telemetry requests.

joshbetz committed:
Add e2e tests for telemetry endpoint

* Verify that missing parameters will result in errors
* Verify that passing correct parameters will result in 200

joshbetz committed:
Additional invalid tests cases

Use it.each to run the same test with various invalid test data

joshbetz committed:
lint

jonathansadowski commented on includes/rest-api/Controllers/Version2/class-wc-rest-telemetry-v2-controller.php:
```diff
@@ -0,0 +1,142 @@
+	protected $rest_base = 'telemetry';
+
+	/**
+	 * Register the route for /system_status
```
Looks like the comment here was carried over from the class used as an example.

jonathansadowski commented on includes/rest-api/Controllers/Version3/class-wc-rest-telemetry-controller.php:
```diff
@@ -0,0 +1,27 @@
+/**
+ * REST API WC Telemetry Controller
+ *
+ * Handles requests to the /system_status/App/* endpoints.
```
This docblock doesn't seem to be accurate.

jonathansadowski reviewed (changes requested):
Sorry for the delay in reviewing this. I commented about some comments/docblocks that seem to be inaccurate / need updating, otherwise this tested well for me.

I do have some inarticulable pause about adding telemetry to the v3 namespace. I do wonder if it makes sense to place it in some other namespace. Placing it there really encourages its usage outside of the apps, in my mind... but I'm not sure if that's just me. /cc @vedanshujain did you have any opinions in that regard?

joshbetz commented:
Thanks @jonathansadowski. I'll double check all the comment blocks and make sure to fix the ones you mentioned.

> I do have some inarticulable pause about adding telemetry to the v3 namespace.

I don't have any strong opinions on this. Happy to put it in a new namespace or even `wc-analytics`.

joshbetz committed:
Fix the permission check for telemetry data

Any logged in user should be able to post telemetry data

joshbetz committed:
Fix the docblock copy pasta

joshbetz committed:
Change endpoint to wc-telemetry/tracker

Since this endpoint isn't a publicly documented endpoint, we'll keep it
out of the wc/v3 namespace. Instead we're introducing a new wc-telemetry
namespace. The name of the endpoint then is updating to tracker.

The wc-telemetry endpoint is unversioned for consistency with
wc-analytics and because it's not publicly documented.

joshbetz committed:
Add docblock for get_telemetry_controllers

jonathansadowski commented on tests/e2e/core-tests/specs/api/telemetry.test.js:
```diff
@@ -0,0 +1,61 @@
+			{ version: '1.1' },
+		])( 'errors for invalid request body - %p', async data => {
+			const response = await client
+				.post( `/wc/v3/telemetry`, data )
```
Looks like this still needs to be updated with the new route

jonathansadowski commented on includes/rest-api/Controllers/Telemetry/class-wc-rest-telemetry-controller.php:
```diff
@@ -0,0 +1,142 @@
+/**
+ * REST API WC Telemetry controller
+ *
+ * Handles requests to the /telemetry endpoint.
```
Can you update this to `/tracker`?

jonathansadowski commented on includes/rest-api/Controllers/Telemetry/class-wc-rest-telemetry-controller.php:
```diff
@@ -0,0 +1,142 @@
+	protected $rest_base = 'tracker';
+
+	/**
+	 * Register the route for /telemetry
```
Same as above

jonathansadowski commented on tests/e2e/core-tests/specs/api/telemetry.test.js:
```diff
@@ -0,0 +1,61 @@
+
+		it( 'returns 200 with correct fields', async () => {
+			const response = await client
+				.post( `/wc/v3/telemetry`, {
```
Same as above

jonathansadowski reviewed (changes requested):
Looking good — I added a few minor comments, but otherwise everything tested out well. Thanks for updating the namespace used!

joshbetz committed:
Update tests to use new endpoint

jonathansadowski reviewed (dismissed):
Thanks for making the requested updates. Everything looks good. I verified that the test failures are unrelated to this change, and that the unit tests pass for me when I run locally with this branch.

jonathansadowski reviewed (approved):
Thanks for the rebase. All tests good except for the sniff re: `Server.php` and the filename, but I don't think we should fix that in the context of this PR.

Looking good, nice work.

jonathansadowski commented:
I didn't add the needs changelog or needs testing instructions labels because of the scope of this API being limited to mobile consumption.
